### PR TITLE
Auto scanner

### PR DESCRIPTION
**Pull request by starwarigami**
Auto scanner functionality

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (SorenHolstHansen) [inline .vscode/launch.json:1]: Delete this file and put it in .gitignore

File: src/icons/AutoScan.svelte [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<svg
```
Comment (SorenHolstHansen) [inline src/icons/AutoScan.svelte:1]: Use phosphor icons

File: src/icons/Pause.svelte [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<svg
```
Comment (SorenHolstHansen) [inline src/icons/Pause.svelte:1]: use Phosphor icons

File: src/icons/Play.svelte [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<svg
```
Comment (SorenHolstHansen) [inline src/icons/Play.svelte:1]: use Phosphor icons

File: src/icons/Scanner.svelte [hunk 5]
```diff
@@ -0,0 +1,66 @@
+<svg
```
Comment (SorenHolstHansen) [inline src/icons/Scanner.svelte:1]: use Phosphor icons if there is an alternative

File: src/icons/Step.svelte [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<svg
```
Comment (SorenHolstHansen) [inline src/icons/Step.svelte:1]: use Phosphor icons

File: src/server/trpc/users.ts [hunk 7]
```diff
@@ -163,4 +163,40 @@ export default trpc
...
+      const jwtToken = getJwt(ctx);
+      if (jwtToken != null) {
+        const settingsRaw = await ctx.prisma.userSettings.findUnique({ where: { userId: jwtToken.id } });
+        if (settingsRaw != null) {
+          const settings = UserSettingsValidator.parse(settingsRaw);
+          if (settings != null) {
+            return settings;
+          }
+        }     
+      }
```
Comment (SorenHolstHansen) [inline src/server/trpc/users.ts:178]:
```suggestion
      if (ctx.token != null) {
        const settingsRaw = await ctx.prisma.userSettings.findUnique({ where: { userId: ctx.token.id } });
        if (settingsRaw != null) {
          const settings = UserSettingsValidator.parse(settingsRaw);
          if (settings != null) {
            return settings;
          }
        }     
      }
```

File: src/server/trpc/users.ts [hunk 8]
```diff
@@ -163,4 +163,40 @@ export default trpc
...
+        await ctx.prisma.userSettings.create({
+          data: { userId: jwtToken.id, ...input }
+        });
+      } catch (e) {
+        
+      await ctx.prisma.userSettings.update({
+        where: { userId: jwtToken.id },
+        data: { userId: jwtToken.id, ...input}
+      });
+      }
```
Comment (SorenHolstHansen) [inline src/server/trpc/users.ts:200]:
```suggestion
      if (ctx.token == null) {
        throw new TRPCError({ message: 'You are not logged in', code: 'UNAUTHORIZED' });
      }
      try {
        await ctx.prisma.userSettings.create({
          data: { userId: ctx.token.id, ...input }
        });
      } catch (e) {
        
      await ctx.prisma.userSettings.update({
        where: { userId: ctx.token.id },
        data: { userId: ctx.token.id, ...input}
      });
      }
```

File: src/stores/sudokuStore/scanner.ts [hunk 9]
```diff
@@ -0,0 +1,801 @@
...
+    getHighlightedCells,
+    toggleSeen,
+    toggleTuples
+  };
+}
+/**
+ * The history of editor steps.
+ * A field is either the current value, or an index of the last time it changed
+ */
+export const scanner = createScannerStore();
```
Comment (SorenHolstHansen) [inline src/stores/sudokuStore/scanner.ts:801]: This file is very long and the functions are difficult to figure out. Is it possible to split into more files, or to split the functions up into smaller ones

**Review by SorenHolstHansen — CHANGES_REQUESTED** [review]

File: src/icons/AutoScan.svelte (hunk 2 above)
Comment (starwarigami) [reply src/icons/AutoScan.svelte]: Done

File: src/icons/Pause.svelte (hunk 3 above)
Comment (starwarigami) [reply src/icons/Pause.svelte]: Done

File: src/icons/Play.svelte (hunk 4 above)
Comment (starwarigami) [reply src/icons/Play.svelte]: Done

File: src/icons/Scanner.svelte (hunk 5 above)
Comment (starwarigami) [reply src/icons/Scanner.svelte]: Done

File: src/icons/Step.svelte (hunk 6 above)
Comment (starwarigami) [reply src/icons/Step.svelte:1]: There is no composite Play/Pause icon, so the custom icon is necessary here

File: .vscode/launch.json (hunk 1 above)
Comment (starwarigami) [reply .vscode/launch.json]: Done

File: src/server/trpc/users.ts (hunk 7 above)
Comment (starwarigami) [reply src/server/trpc/users.ts:178]: Taken suggested code

**Comment by SorenHolstHansen** [comment]
There are some merge conflicts, I will rereview after those are resolved :)

File: src/icons/Scanner.svelte [hunk 10]
```diff
@@ -1,66 +1,5 @@
...
-    stroke-linecap="round"
-    stroke-linejoin="round"
-    stroke-width="16"
-  /></svg
->
+<script>
+  import Crosshair from 'phosphor-svelte/lib/Crosshair/Crosshair.svelte';
+</script>
+
+<Crosshair size={64} />
```
Comment (SorenHolstHansen) [inline src/icons/Scanner.svelte:5]: I think you should inline this component wherever you are using it instead of making a wrapper component

**Review by SorenHolstHansen — DISMISSED** [review]
Looks really good. There is just the one thing which I wrote below, otherwise you are good to merge after you have fixed merge conflicts

**Comment by starwarigami** [comment]
The wrapper is so that I can override the default size of the phosphor
icon. The Controller requires all button icons to be the NumbersIcon type,
which doesn't have a size property

**Review by SorenHolstHansen — APPROVED** [review]

**Pull request merged**